### PR TITLE
Fix issue 682: Make history, as-of and since composable with each other

### LINKED ISSUE #682: Make history, as-of and since composable
### Describe the feature you would like to request

Currently the following three queries all produce an error:

```clj
(d/q '[:find ?a :in $ ?e :where [?e :age ?a]]
                        (-> @conn d/history d/history)
                        [:name "Alice"])

(d/q '[:find ?a :in $ ?e :where [?e :age ?a]]
                    (-> @conn
                        (d/as-of tx-id0)
                        d/history)
                    [:name "Alice"])

(d/q '[:find ?a :in $ ?e :where [?e :age ?a]]
                    (-> @conn
                        d/history
                        (d/as-of tx-id0))
                    [:name "Alice"])
```

The error is the following:

```
error: java.lang.IllegalArgumentException: No implementation of method: :-slice of protocol: #'datahike.index.interface/IIndex found for class: nil
```

### Describe the solution you would like

I would like to make queries against database expressions such as `(-> @conn d/history d/history)`, `(-> @conn (d/as-of tx-id0) d/history)` where we compose calls to `d/history`, `d/as-of` and `d/since`.

### Describe alternatives you've considered

Datomic, where this is possible (at least to some extent).

### PR DESCRIPTION
**Pull request by jonasseglare**
#### SUMMARY

Fixes #682 

Makes it possible to compose `history`, `as-of` and `since` in arbitrary ways, e.g. `(-> @conn history (as-of tx-id) (since tx-id2))`.

##### Feature
- [ ] Implements an existing feature request. Make sure the feature request has been accepted for implementation before opening a PR.
- [x] Related issues linked using `fixes #number`
- [x] Integration tests added
- [ ] Documentation added
- [ ] Architecture Decision Record added 
- [ ] Formatting checked

### CODE REVIEW

File: src/datahike/db.cljc [hunk 1]
```diff
@@ -137,6 +137,68 @@
...
+
+(defn deeper-index-range [origin-db db attr start end context]
+  (dbi/-index-range origin-db
+                    attr
+                    start
+                    end
+                    (merge {:current-db db}
+                           context)))
+
+(defmacro todo [x] x)
```
Comment (whilo) [inline src/datahike/db.cljc:199]: Is this a leftover?

**Review by whilo — CHANGES_REQUESTED** [review]
Overall this looks good. I think not all of these compositions make sense, but it is nice if it works this way.

File: src/datahike/db.cljc (hunk 1 above)
Comment (jonasseglare) [reply src/datahike/db.cljc]: Yes, thanks for spotting that. I have removed it.

**Review by whilo — APPROVED** [review]
That makes sense, thanks for providing this improved compatibility with Datomic.

**Pull request merged**